lib/mongoid/extensions/regexp.rb

```diff
@@ -12,12 +12,18 @@ module ClassMethods
         # @example Mongoize the object.
         #   Regexp.mongoize(/\A[abc]/)
         #
-        # @param [ Regexp, String ] object The object to mongoize.
+        # @param [ Object ] object The object to mongoize.
         #
-        # @return [ Regexp ] The object mongoized.
+        # @return [ Regexp | nil ] The object mongoized or nil.
         def mongoize(object)
-          return nil if object.nil?
-          ::Regexp.new(object)
+          return if object.nil?
+          case object
+          when String then ::Regexp.new(object)
+          when ::Regexp then object
+          when BSON::Regexp::Raw then object.compile
```

Comment:
This is a lossy operation I'm afraid:

```

irb(main):014:0> a=BSON::Regexp::Raw.new('abc','m')
=> #<BSON::Regexp::Raw:0x00007fbd22b28e00 @options="m", @pattern="abc">
irb(main):015:0> a.compile.options
=> 0

```

The declared field type is a Ruby regexp though, not bson one.

What I think can be done is a warning here stating that BSON::Regexp::Raw conversion to Ruby Regexp is lossy, ideally one time per program execution.


Alternatively incoming value of BSON::Regexp::Raw could be  returned as is from `mongoize` and written exactly as is into attributes, with  `demongoize`  issuing the `compile` call to get the Ruby regexp, but conceptually this behavior doesn't align with what I would expect from a field declared as a Ruby regexp.